src/curve_intersections.rs

```diff
@@ -0,0 +1,604 @@
+                intersections,
+                !flip,
+                recursion_count,
+                call_count,
+                orig_curve2,
+                orig_curve1,
+            );
+        }
+    } else {
+        // Iterate.
+        if domain2.end - domain2.start >= epsilon {
+            call_count = add_curve_intersections(
+                curve2,
+                curve1,
+                domain2,
+                new_domain1,
+                intersections,
+                !flip,
+                recursion_count,
+                call_count,
+                orig_curve2,
+                orig_curve1,
+            );
+        } else {
+            // The interval on curve2 is already tight enough, so just continue iterating on curve1.
+            call_count = add_curve_intersections(
+                curve1,
+                curve2,
+                new_domain1,
+                domain2,
+                intersections,
+                flip,
+                recursion_count,
+                call_count,
+                orig_curve1,
+                orig_curve2,
+            );
+        }
+    }
+
+    call_count
+}
+
+fn add_point_curve_intersection<T: ParamCurveBezierClipping, U: ParamCurveBezierClipping>(
+    pt_curve: &T,
+    pt_curve_is_curve1: bool,
+    curve: &U,
+    pt_domain: &Range<f64>,
+    curve_domain: &Range<f64>,
+    intersections: &mut ArrayVec<[(f64, f64); 9]>,
+    flip: bool,
+) {
+    let pt = pt_curve.start();
+    // We assume pt is curve1 when we add intersections below.
+    let flip = if pt_curve_is_curve1 { flip } else { !flip };
+
+    // Generally speaking |curve| will be quite small at this point, so see if we can get away with
+    // just sampling here.
+
+    let epsilon = epsilon_for_point(pt);
+    let pt_t = (pt_domain.start + pt_domain.end) * 0.5;
+
+    let curve_t = {
+        let mut t_for_min = 0.0;
+        let mut min_dist_sq = epsilon;
+        let tenths = [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0];
+        for &t in tenths.iter() {
+            let d = (pt - curve.eval(t)).hypot2();
+            if d < min_dist_sq {
+                t_for_min = t;
+                min_dist_sq = d;
+            }
+        }
+
+        if (min_dist_sq - epsilon).abs() < f64::EPSILON {
+            -1.0
+        } else {
+            domain_value_at_t(curve_domain, t_for_min)
+        }
+    };
+
+    if (curve_t - -1.0).abs() > f64::EPSILON {
+        add_intersection(pt_t, pt_curve, curve_t, curve, flip, intersections);
+        return;
+    }
+
+    // If sampling didn't work, try a different approach.
+    let results = point_curve_intersections(pt, curve, epsilon);
+    for t in results {
+        let curve_t = domain_value_at_t(curve_domain, t);
+        add_intersection(pt_t, pt_curve, curve_t, curve, flip, intersections);
+    }
+}
+
+fn point_curve_intersections<T: ParamCurveBezierClipping>(
+    pt: Point,
+    curve: &T,
+    epsilon: f64,
+) -> ArrayVec<[f64; 9]> {
```

Comment:
If you don't mind, I think I'll have a crack at implementing (and testing) this. I'll try to make the function as general as possible so that it can be useful to other parts of the code as well. When I'm done I'll pull request it into your branch so that you can review it.